clippy_lints/src/unsafe_sizeof_count_copies.rs

```diff
@@ -0,0 +1,97 @@
+    /// # use std::mem::size_of;
+    ///
+    /// const SIZE: usize = 128;
+    /// let x = [2u8; SIZE];
+    /// let y = [2u8; SIZE];
+    /// unsafe { copy_nonoverlapping(x.as_ptr(), y.as_mut_ptr(), size_of::<u8>() * SIZE) };
+    /// ```
+    pub UNSAFE_SIZEOF_COUNT_COPIES,
+    correctness,
+    "unsafe memory copying using a byte count instead of a count of T"
+}
+
+declare_lint_pass!(UnsafeSizeofCountCopies => [UNSAFE_SIZEOF_COUNT_COPIES]);
+
+fn get_size_of_ty(cx: &LateContext<'tcx>, expr: &'tcx Expr<'_>) -> Option<TyM<'tcx>> {
+    match &expr.kind {
+        ExprKind::Call(ref count_func, func_args) => {
+            if_chain! {
+                if let ExprKind::Path(ref count_func_qpath) = count_func.kind;
+                if let Some(def_id) = cx.qpath_res(count_func_qpath, count_func.hir_id).opt_def_id();
+                if match_def_path(cx, def_id, &paths::MEM_SIZE_OF)
+                    || match_def_path(cx, def_id, &paths::MEM_SIZE_OF_VAL);
+                then {
+                    cx.typeck_results().node_substs(count_func.hir_id).types().next()
+                } else {
+                    None
+                }
+            }
+        },
+        ExprKind::Binary(op, left, right) if &BinOpKind::Mul == &op.node || &BinOpKind::Div == &op.node => {
+            get_size_of_ty(cx, &*left).or_else(|| get_size_of_ty(cx, &*right))
+        },
+        _ => None,
+    }
+}
+
+impl<'tcx> LateLintPass<'tcx> for UnsafeSizeofCountCopies {
+    fn check_expr(&mut self, cx: &LateContext<'tcx>, expr: &'tcx Expr<'_>) {
+        if_chain! {
+            // Find calls to ptr::copy and copy_nonoverlapping
+            if let ExprKind::Call(ref func, ref func_args) = expr.kind;
+            if let ExprKind::Path(ref func_qpath) = func.kind;
+            if let Some(def_id) = cx.qpath_res(func_qpath, func.hir_id).opt_def_id();
+            if match_def_path(cx, def_id, &paths::COPY_NONOVERLAPPING)
+                || match_def_path(cx, def_id, &paths::COPY);
+
+            // Get the pointee type
+            let substs = cx.typeck_results().node_substs(func.hir_id);
+            if let Some(pointee_ty) = cx.typeck_results().node_substs(func.hir_id).types().next();
+
+            // Find a size_of call in the count parameter expression and
+            // check that it's the same type
+            if let [src, dest, count] = &**func_args;
+            if let Some(ty_used_for_size_of) = get_size_of_ty(cx, count);
+            if TyS::same_type(pointee_ty, ty_used_for_size_of);
+            then {
+                span_lint_and_help(
+                    cx,
+                    UNSAFE_SIZEOF_COUNT_COPIES,
+                    expr.span,
+                    "unsafe memory copying using a byte count (Multiplied by size_of::<T>) instead of a count of T",
+                    None,
+                    "use a count of elements instead of a count of bytes for the count parameter, it already gets multiplied by the size of the pointed to type"
```

Comment:
It's definitely weird, that the test fails, but `clippy dev fmt` passes. But the reason the test fails is that those lines are too long and rustfmt can't handle complicated macros.

Reply:
I tried making them multiline strings, and even removing the entire span_lint_and_help call, and I still get the same error. It looks like https://github.com/rust-lang/rustup/issues/1137

Reply:
Ah, I can't really help you on windows. You can try to get it through CI.

Reply:
Ok! I think it wasn't running because of the conflict on the removed lintlist module (The new lint script added something there). I just rebased, deleted that file in my commit and force pushed, it's running now. 🤞 

I tried rerunning `rustup-toolchain-install-master` to see if that fixed it, but it seems rustfmt might not be available on the latest nightly